### ISSUE TITLE
Debug.trace in test leads to cryptic and inconsistent error

### ISSUE THREAD

**Issue by ceedubs** [unisonweb/unison#4685]
# Debug.trace in test leads to cryptic and inconsistent errors

If you have a `Debug.trace` call in a test, then a number of confusing things will happen:

- ucm will bug out with `pure code can't perform I/O`. Since tests are generally cached, this may be a reasonable thing to do, but people might not think of `Debug.trace` as IO, so it might be helpful to point out that it can be the cause of this error. It gets worse from here.
- **Edit: this point is partially fixed by [#5585](https://github.com/unisonweb/unison/pull/5585) to include the test name, but not the violation which may be very nested.**  It doesn't include the offending test in the test results, so you can't even tell that this test was run. A bug message prints to the screen but it has a stack trace that just has `builtin.bug` and a hash that doesn't actually correspond to the hash of the test term.

      Stack trace:
        builtin.bug
        #oit1slb1i3
- It's inconsistent because the IO sandbox doesn't apply to `test>` watch expressions in your scratch file.
- It's really easy to miss the introduction of the `Debug.trace`, because once it passes with the `test>` watch expression, the result is cached and you won't hit the sandbox error. Someone _else_ will hit it when they pull your changes and run `test` without having a cached result for it.

## Raw Transcript

<details>
<summary>Expand to copy raw transcript</summary>

````
```ucm:hide
.> builtins.merge
```

This passes, because the IO sandbox doesn't seem to apply to `test>` watch expressions.

```unison
test> foo.test =
  x = 192
  Debug.trace "x" x
  [Ok "Passed"]
```

```ucm
.> add
```

The `test` command succeeds, because the result of `foo.test` is already cached, which skips the IO sandbox.

```ucm
.> test
```

This test which is essentially the same will fail, because it is never run with a `test>` watch expression to get the result into the test cache.

```unison
bar.test : [Test.Result]
bar.test =
  x = 42
  Debug.trace "x" x
  [Ok "Passed"]
```

```ucm
.> add
.> test
```
````

</details>

## Transcript output

This passes, because the IO sandbox doesn't seem to apply to `test>` watch expressions.

```unison
test> foo.test =
  x = 192
  Debug.trace "x" x
  [Ok "Passed"]
```

```ucm

  Loading changes detected in scratch.u.

  I found and typechecked these definitions in scratch.u. If you
  do an `add` or `update`, here's how your codebase would
  change:
  
    ⍟ These new definitions are ok to `add`:
    
      foo.test : [Result]
  
  Now evaluating any watch expressions (lines starting with
  `>`)... Ctrl+C cancels.

    2 |   x = 192
    
    ✅ Passed Passed

```
```ucm
.> add

  ⍟ I've added these definitions:
  
    foo.test : [Result]

```
The `test` command succeeds, because the result of `foo.test` is already cached, which skips the IO sandbox.

```ucm
.> test

  Cached test results (`help testcache` to learn more)
  
  ◉ foo.test   Passed
  
  ✅ 1 test(s) passing
  
  Tip: Use view foo.test to view the source of a test.

```
This test which is essentially the same will fail, because it is never run with a `test>` watch expression to get the result into the test cache.

```unison
bar.test : [Test.Result]
bar.test =
  x = 42
  Debug.trace "x" x
  [Ok "Passed"]
```

```ucm

  Loading changes detected in scratch.u.

  I found and typechecked these definitions in scratch.u. If you
  do an `add` or `update`, here's how your codebase would
  change:
  
    ⍟ These new definitions are ok to `add`:
    
      bar.test : [Result]

```
```ucm
.> add

  ⍟ I've added these definitions:
  
    bar.test : [Result]

.> test

    
    Cached test results (`help testcache` to learn more)
    
    ◉ foo.test   Passed
    
    ✅ 1 test(s) passing
    
    ✅  

  

  💔💥
  
  I've encountered a call to builtin.bug with the following
  value:
  
    "pure code can't perform I/O"
  
  
  Stack trace:
    builtin.bug
    #oit1slb1i3

```

🛑

The transcript failed due to an error in the stanza above. The error is:

  💔💥
  
  I've encountered a call to builtin.bug with the following
  value:
  
    "pure code can't perform I/O"
  
  
  Stack trace:
    builtin.bug
    #oit1slb1i3

**Comment by aryairani** [comment]
I ran into this also.

```
Debug.trace "..."
```
fails

```
when false do Debug.trace "..."
```
fails

```
_ = do Debug.trace "..."
```
passes.

One repro with
```
clone @aryairani/hnsw-anns/@aryairani/pure-code-cant-perform-io
tests
```

**Comment by ceedubs** [comment]
This still happens and is super bewildering, because the test that it complains about doesn't show up anywhere in the output.

**Comment by aryairani** [comment]
What should it do instead?

Option 1: make it bomb in `test>` too.
Option 2: make it not bomb when it's run from the branch, although the trace messages aren't cached.
Option 3: save and reprint the trace messages, which are deterministic, when you view the cached result.

maybe there's a `{Debug.Trace}` ability that collects them, and UCM could know about it, .

```
ability Debug where trace : Text -> a -> ()
type Debug.Traced a = Traced [(Text,Text)] a
handleTrace : abilities.Request {Debug} a -> Traced a
handleTrace = 
  loop z = cases
    { trace t a -> k } -> 
      handle 
        ##Debug.trace t a
        k()
      with 
        loop (z :+ (t, toDebugText a))
    { a } -> Traced z a
  loop []
```

except we don't want traces to show up in the type, so I guess it can't be exactly this.

Or maybe because we only care about `test>` we could do some automated change when it's run?

---

Can we think of any reason not to do Option 2?

**Comment by ceedubs** [comment]
While perhaps the most technically correct thing to do, I think that option `1` is hostile to development/debugging and am not inclined to go with it, at least for the case of `Debug.trace`.

I was about to agree that we should just go with option `2`, but I remembered an experience that Fabio and I had that makes me question this approach. In our case he had a function that referenced an IO operation without actually calling it. This worked fine in a `test>` block but failed in `test` if you didn't already have a cached result. So I think that we'd either need to change the behavior to only limit _running_ sandboxed operations and not just referencing them; OR we might want to change `test>` to prohibit referencing sandboxed ops (but possibly add an exception to allow `Debug.trace`). I do think that there might be some precedent with the `Doc` renderer to allow referencing sandboxed operations but use a runtime that bombs if they are run.

**Comment by aryairani** [comment]
@dolio Could we get your take on this too

**Comment by dolio** [comment]
It seems like we should be able to make tests use the sandboxed _runtime_ instead of what it's doing now. This only dies when you actually run a sandboxed operation.

However, `Debug.trace` is on the list of operations you're not allowed to call in that. We could remove it from the banned operations, because it's relatively harmless. Then you might end up with tests where people forgot to remove the debugging, though.

I think trying to make the cached values generate debug output is more work than it's worth. It's not really obvious how to even do that.

**Comment by ceedubs** [comment]
I agree that trying to make cached values generate debug output is more work than it's worth.

I like @dolio's suggestion of making tests use the sandboxed runtime as opposed to a sandboxed dependency check. I'd be inclined to allow `Debug.trace` within the sandboxed runtime; its usefulness while debugging is worth the potential cost of checking in tests with debug output. If we ever add support for running tests on Share, it might want to use a runtime in which `Debug.trace` is a noop, but I don't think that it would need to flag it.

**Comment by dolio** [comment]
Actually, I could make `Debug.trace` just a no-op in the sandboxed runtime. Then when you're running `test>` watches,  you'd see the output, but it wouldn't show up when running tests, even if it's in there. And then also docs with `Debug.trace` calls wouldn't actually print anything (but wouldn't bomb).

We need to decide one thing, though. The current test runner allows `toDebugText` and `value` to be called. I think the latter is already allowed in the sandboxed runtime, but the former isn't. So, we'd need to allow it. I think it's probably fine, though.

**Comment by ceedubs** [comment]
Ooh allowing `Debug.trace` in a `test>` expression but making it a noop when running `test` sounds like the best of both worlds. I didn't think that was an option.

On the one hand, I think that there is a pretty strong argument against allowing `toDebugText` in cached tests that are supposed to be deterministic. If you write code that calls `toDebugText` on a value and inspects the result then the behavior of that code can change with a definition rename, even if the code hashes don't change. However, being able to call `toDebugText` is kind of helpful when creating test suites and rendering results of failed tests back to the user, so I'm still a bit inclined to allow it. The `test.verify` function in base (which is used fairly heavily) depends on it, for example.
